### ISSUE TITLE
OJS3 Accessibility Improvements

### ISSUE THREAD

**Issue by sean-xiao-zhao7** [pkp/pkp-lib#4150]
Please fix accessibility errors found throughout the OJS3 app by accessibilty checker such as https://achecker.ca/checker/index.php

It'd be great if the entire OJS3 app can pass the WCAG 2.0 / AA level of accessibility.

Here is the error report of AChecker that was ran on a journal using the manuscript child theme:
[achecker_2018-10-04_00-05-36.pdf](https://github.com/pkp/pkp-lib/files/2484437/achecker_2018-10-04_00-05-36.pdf)

**Comment by asmecher** [comment]
@sean-xiao-zhao7, we could address those warnings fairly simply by the book, but that would lead to no improvement to the accessibility of the page. The missing labels are addressed by `aria-label` attributes, for example.

We've run exercises with accessibility checkers before but my read is that the checkers are not a good substitute for expert, experiential testing, and that using the checkers as primary tools for improving accessibility doesn't get good results.

We would love to partner with someone who has accessibility expertise and in particular a group with screen reader experience. So far, we've heard anecdotally that OJS 3.x is "pretty good", but that's not much to go by; our team unfortunately doesn't have the specific accessibility experience to assess in depth.

**Comment by sean-xiao-zhao7** [comment]
Thank you, Alec, completely agreed – this was along the lines of what we have replied to our journals concerned with web accessibility. 

Unfortunately they do tend to go “by the book” and get upset about AChecker / Wave warnings. So if there is an easy fix for those, even with no factual improvement, it will bring some peace of mind. Thank you!

**Comment by jmacgreg** [comment]
To add to @asmecher's comments: PKP is looking at writing up a [VPAT](https://en.wikipedia.org/wiki/Voluntary_Product_Accessibility_Template) for OJS/OMP 3, which will require a much more involved review of the system than what online accessibility tools typically provide. We're a small team though, and would welcome any experienced community interest in assisting with this endeavour. If you or anyone else has experience in this area and have the time to contribute, please do let us know here or via DM. 

Thanks!
James

**Comment by asmecher** [comment]
Tagging @NateWr on those particular complaints as he's worked the most with front-end/template code -- but it'll be pretty hard for us to prioritize changes to accessibility code that don't actually make improvements. I think they're actually damaging to accessibility; they present the illusion of accessibility without actually delivering it, and make it easier for people to claim compliance without delivering it.

**Comment by NateWr** [comment]
Thanks for this, @sean-xiao-zhao7. As others have said, the automated checks have mixed reliability. However, I understand that you've got clients you need to respond to. Here's a quick run-down of the issues raised in the PDF you posted.

If I say a report is "inaccurate", I mean that the reporting tool is wrong. There is no WCAG 2.0 non-compliance.

If I say a report is a "notice", I mean that the reporting tool is telling us to take a closer look but does not necessarily mean WCAG 2.0 non-compliance. Further details below.

- Page 1 - 1.3 - These reports are inaccurate.
- Page 1 - 2.4 - 2.4.4 - This is irrelevant. `aria-hidden` is applied to remove this from a screen-reader's interaction and an alternate control is provided.
- Page 1 - 2.4 - 2.4.6 - Are you able to determine which page this occurred on (an article page, an issue page, etc)? I'm not able to reproduce the heading order issue, though I wouldn't be surprised if one does exist on some page.
- Page 2 - 3.3 - These reports are inaccurate.
- Page 4 - 1.1 - As far as I can tell, these are notices and not errors. But it is you or your joural's responsibility to provide the correct alt text for `https://jps.library.utoronto.ca/public/site/pageHeaderTitleImage_en_US.png`. This can be done under Settings > Website > Appearance and Administration > Site Settings. The other image identified, `https://jps.library.utoronto.ca/templates/images/ojs_brand.pnghttps://jps.library.utoronto.ca/templates/images/ojs_brand.png`, has appropriate alt text.
- Page 4/5 - 1.3 - These reports are inaccurate.
- Page 5/6 - 1.4 - 1.4.1 - These reports appear to be notices, not errors, and none of the images indicated seem to contravene the rules mentioned.
- Page 6 - 1.4 - 1.4.5 - The first image, `https://jps.library.utoronto.ca/public/site/pageHeaderTitleImage_en_US.png`, probably needs a better alt text, which you can provide as described above. The other image mentioned here has appropriate alt text.
- Page 6 - 2.1/2.3 - These reports appear to be notices, not errors. I am not aware of any issues related to keyboard accessibility or flashing, but happy to correct any if found.
- Page 7 - 2.4 - 2.4.1 - These reports are inaccurate.
- Page 7 - 2.4 - 2.4.2 - This report appears to be a notice, not an error. The title from your journal looks :+1:.
- Page 7-10 - 2.4 - 2.4.4 - These appear to be notices that are raised arbitrarily on _every single link_. There is no error here.
- Page 10 - 2.4 - 2.4.5 - We are WCAG 2.0 compliant because we provide two of the [recognized navigation options](https://www.w3.org/WAI/WCAG21/quickref/?versions=2.0&showtechniques=245#multiple-ways) (links and search). A sitemap would be a nice addition, but it's not necessary for WCAG 2.0 compliance. I've filed an issue: https://github.com/pkp/pkp-lib/issues/4153
- Page 10 - 2.4 - 2.4.6 - These are notices, not errors, and are incorrect.
- Page 11 - 3.1 - This is a notice, not an error.
- Page 11 - 3.2 - 3.2.3 - As far as I'm aware, there is no issue here. This rule appears to rely on a judgement of "predictable". Happy to discuss further.
- Page 11 - 3.2 - 3.2.4 - This report would related to specific content on your site. I wasn't able to find any offending content on the linked site.
- Page 12 - 3.3 - These reports are either inaccurate or are just notices, which do not seem to apply in these cases.

I don't see anything that we need to change in the theme's code here. I'm going to close this issue but feel free to reply with any questions/concerns.

**Comment by sean-xiao-zhao7** [comment]
thanks @NateWr for the detailed analysis, these are very helpful. And @asmecher @jmacgreg for the help
